Draft of webactions

nancyhild opened the pull request:
This file was very straightfoward but there are some places where I wasn't sure how/whether to change what was there--either references to whisk or links to other OW articles. Please search for [[NH:  to see those spots.

nancyhild committed:
WIP checkin

Committing progress to date.

nancyhild committed:
Interim save of actions.md

Done through activation section.

nancyhild committed:
Update actions.md

Work in progress.

nancyhild committed:
Ready for PR

Not polished but needs review.

nancyhild committed:
Removed license

Just beginning.

nancyhild committed:
Update actions.md

Trying again

nancyhild committed:
Edited for nim

joshuaauerbachwatson commented on webactions.md:
```diff
@@ -0,0 +1,546 @@
+
+The default `content-type` for an HTTP response is `application/json` and the body can be any allowed JSON value. The default `content-type` can be omitted from the headers.
+
+**[[NH: Note link to reference.md]]**
```
I'm asking @rabbah to comment on which linked documents should be in scope for your translation effort versus which should be dropped.  He knows what's in the other documents and whether or not translation is worth it.

rabbah replied:
we will need our own reference.md to document the limits - starting with the apache version is ok.

nancyhild replied:
ok modified the sentence a bit so people are not surprised when they click the link.

joshuaauerbachwatson commented on webactions.md:
```diff
@@ -0,0 +1,546 @@
+
+## Securing web actions
+
+**[[NH: I don't know what to do with this section, please advise.]]**
```
Mostly, the section is still applicable substituting `nim` for `wsk`.   I'll note exceptions.

joshuaauerbachwatson commented on webactions.md:
```diff
@@ -0,0 +1,546 @@
+**[[NH: I don't know what to do with this section, please advise.]]**
+By default, a web action can be invoked by anyone having the web action's invocation URL. Use the `require-whisk-auth` [web action annotation](annotations.md#annotations-specific-to-web-actions) to secure the web action. When the `require-whisk-auth` annotation is set to `true`, the action will authenticate the invocation request's Basic Authorization credentials to confirm they represent a valid ...
+
+Alternatively, use the `--web-secure` flag to automatically set the `require-whisk-auth` annotation.  When set to `true` a random number is generated as the `require-whisk-auth` annotation value. When set to `false` the `require-whisk-auth` annotation is removed.  When set to any other value, that value is used as the `require-whisk-auth` annotation value.
```
The `nim` command supports `--web-secure` but there is a difference.  A value of `true` just sets the annotation to `true`.   It does not generate a random number.

nancyhild replied:
Rewrote this section and added subsections.

joshuaauerbachwatson commented on webactions.md:
```diff
@@ -0,0 +1,546 @@
+
+## Vanity URL
+
+Web actions may be accessed via an alternate URL, which treats the Nimbella namespace as a subdomain to the API host. This is suitable for developing web actions that use cookies or local storage so that data is not inadvertently made visible to other web actions in other namespaces. The namespaces must match the regular expression `[a-zA-Z0-9-]+` (and should be 63 characters or fewer) for the ed ...
```
We have similar functionality in `nim` but not the same in detail.   To avoid confusion, the section needs to be more or less completely replaced with a section about how Nimbella generates URLs for customer namespaces.   I think either I or @rabbah needs to write the section.   You should probably just replace the section with a note to that effect.

rabbah replied:
agree Josh or I will take it.

nancyhild replied:
Note added.

joshuaauerbachwatson reviewed (commented):
I have responded to your specific questions and left indications of where I think @rabbah needs to be consulted.   I have not read the entire text over for correctness.

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+      - the action exceeded its time limit during the init or run phase
+      - the action specified a wrong docker container name
+      - the action did not properly implement the expected [runtime protocol](actions-new.md)
+    - *"whisk internal error"*: the system was unable to invoke the action. **[[NH: is 'whisk' ok here? Note also in table below]]**
```
I think it's a question of what text our system actually uses (the word 'whisk' is not poisonous in itself).  @rabbah should comment.

rabbah replied:
thinking if we want to change this description to something else - for now leave it as is.

joshuaauerbachwatson commented on actions.md:
````diff
@@ -0,0 +1,547 @@
+**[[NH: check output, looks like the original had no package so my command is wrong?]]**
+```
+nim action get greeting
+ok: got action greeting
````
`nim` does not put a tag line like this; currently, the output starts with the JSON

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+Metadata that describes existing actions can be retrieved via the `nim action get` command.
+
+**[[NH: check output, looks like the original had no package so my command is wrong?]]**
```
Do you have `nim` installed locally?   If you can execute it then cut and paste its actual output that will help; often the output of `nim` does look somewhat different than the output of `wsk`, especially when you consider the differences between our backend and the vanilla Apache backend.

nancyhild replied:
No we'll talk about that.

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+1. Save action code to a filename that corresponds with an existing action name in the current working directory. A file extension that corresponds to the action kind is used. An extension of _.zip_ is used for action code that is a zip file.
+
+  **[[NH: please check command and output:]]**
```
`nim` does not display any command line output here (it just returns, having saved the code as you requested)

On the general question of what to do about the fact that we do not provide a /whisk-system/samples package for customers to experiment with, I will let @rabbah comment.

rabbah replied:
these samples are useful for primiarly one thing, saving the user from creating boiler plate functions for the sake of the examples in the docs.

(this once led to the idea to inline functions on the command line :)
if we had a way to generate samples from the command line (nim project create? this would not necessary i think - in the overall doc flow which introduces projects before individual actions, this may work).

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+Here, we see actions listed in order from most to least recently updated. For easier browsing, you can use the flag `--name-sort` or `-n` to sort the list alphabetically:
+
+**[[NH: Output list seems backwards for alphabetical order?]]**
```
Output format has changed recently (don't worry about that ... I think it is not yet in the version of `nim` that you can download and install).  I will be rectifying that soon, but, I see no problem with `--name-sort` when I try it (on a different namespace).   Is this something you executed or is it copied from the Apache document?

nancyhild replied:
It came over from the Apache doc.

joshuaauerbachwatson commented on actions.md:
````diff
@@ -0,0 +1,547 @@
+  /guest/greeting                    private nodejs:6
+```
+
+The list is now sorted alphabetically by namespace, then package name if any, and finally action name, with the default package (no specified package) listed at the top. **[[NH: Is namespace listed in this output example? What does output look like if there's a package? Does nim output list the namespace?]]**
````
For your reference, this is what our output looks like in the latest version (not yet fully public).   I am working to update our public release so you will have the latest code to work with.
```
 nim action list --name-sort
Datetime          Version   Access      Kind           Actions                                           
04/01/20 10:04:60 0.0.5     web open    nodejs:10      /wbtestni-grinjpsjnuh/action                      
03/31/20 14:03:26 0.0.4     web open    nodejs:10      /wbtestni-grinjpsjnuh/billing/awsbill             
03/21/20 16:03:10 0.0.6     web open    php:7.3        /wbtestni-grinjpsjnuh/visits/counter              
03/31/20 14:03:30 0.0.4     web open    nodejs:10      /wbtestni-grinjpsjnuh/billing/datadogbill         
03/31/20 14:03:31 0.0.4     web open    nodejs:10      /wbtestni-grinjpsjnuh/billing/dobill              
03/22/20 11:03:44 0.0.8     private     nodejs:10      /wbtestni-grinjpsjnuh/echo/echo                   
03/23/20 13:03:76 0.0.8     private     php:7.3        /wbtestni-grinjpsjnuh/visits/info                 ```

rabbah replied:
note ive revised this even further but think i am done with those changes.

nancyhild replied:
Copied this in for now, bretter than what was there.

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+To filter your list of actions to just those within a specific package, use this command:
+
+**[[NH: please check command and output:]]**
```
For now I am not checking individual command outputs ... instead will work to get so you can run this everything for yourself.   Worst case (if we're unable to accommodate that), you leave place holders for the output and one of us will fill it in on a later pass.

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+## Access action metadata within the action body
+
+**[[NH: "OpenWhisk assets" in next paragraph ok?]]**
```
OpenWhisk in general is ok.  The Nimbella information model is a superset of the Apache OpenWhisk information model, a fact that we sometimes even highlight.

joshuaauerbachwatson commented on actions.md:
```diff
@@ -0,0 +1,547 @@
+
+The action environment contains several properties that are specific to the running action. These allow the action to programmatically work with OpenWhisk assets via the REST API or set an internal alarm when the action is about to use up its allotted time budget.
+
+**[[NH: Not sure what to do with these properties--please advise. Also note link to annotations.md]]**
```
@rabbah should take this one

rabbah replied:
we have to keep them as is for now.

joshuaauerbachwatson reviewed (commented):
I am belatedly also responding to your queries in actions.md.   I owe you a careful read of it (not yet), since I asked you to change the initial framing and stopped reading at that point.   Hopefully soon.

nancyhild committed:
Update webactions.md

Addressing comments in PR#2

nancyhild committed:
Update links to OpenWhisk docs

Full URLs, mostly to annotations.md and mentioned in the text that these are OpenWhisk docs.

nancyhild committed:
Addressed PR #2 conversations

Need to check all nim output, we'll discuss.

nancyhild committed:
Formatting cleanup
